File: kirun-java/src/main/java/com/fincity/nocode/kirun/engine/function/system/date/DifferenceOfTimeStamps.java
package com.fincity.nocode.kirun.engine.function.system.date;

import java.time.ZoneOffset;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.util.List;
import java.util.Map;

import com.fincity.nocode.kirun.engine.exception.KIRuntimeException;
import com.fincity.nocode.kirun.engine.function.reactive.AbstractReactiveFunction;
import com.fincity.nocode.kirun.engine.json.schema.Schema;
import com.fincity.nocode.kirun.engine.model.Event;
import com.fincity.nocode.kirun.engine.model.EventResult;
import com.fincity.nocode.kirun.engine.model.FunctionOutput;
import com.fincity.nocode.kirun.engine.model.FunctionSignature;
import com.fincity.nocode.kirun.engine.model.Parameter;
import com.fincity.nocode.kirun.engine.namespaces.Namespaces;
import com.fincity.nocode.kirun.engine.runtime.reactive.ReactiveFunctionExecutionParameters;
import com.fincity.nocode.kirun.engine.util.date.IsValidIsoDateTime;
import com.google.gson.JsonPrimitive;

import reactor.core.publisher.Mono;

public class DifferenceOfTimeStamps extends AbstractReactiveFunction {

    private static final String ISO_DATE1 = "dateone";

    private static final String ISO_DATE2 = "datetwo";

    private static final String OUTPUT = "difference";

    private static final Schema dateSchema = Schema.ofRef(Namespaces.DATE + ".timeStamp");

    @Override
    public FunctionSignature getSignature() {

        return new FunctionSignature().setName("DifferenceOfTimeStamps")
                .setNamespace(Namespaces.DATE)
                .setParameters(Map.ofEntries(
                        Parameter.ofEntry(ISO_DATE1, dateSchema),
                        Parameter.ofEntry(ISO_DATE2, dateSchema)))
                .setEvents(Map.ofEntries(
                        Event.outputEventMapEntry(Map.of(OUTPUT, Schema.ofLong(OUTPUT)))));
    }

    @Override
    protected Mono<FunctionOutput> internalExecute(ReactiveFunctionExecutionParameters context) {

        String firstDate = context.getArguments().get(ISO_DATE1).getAsString();

        String secondDate = context.getArguments().get(ISO_DATE2).getAsString();

        if (!IsValidIsoDateTime.checkValidity(firstDate))
            throw new KIRuntimeException("Please provide the valid ISO date for " + ISO_DATE1);

        if (!IsValidIsoDateTime.checkValidity(secondDate))
            throw new KIRuntimeException("Please provide the valid ISO date for " + ISO_DATE2);

        DateTimeFormatter parser = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSS[xx][XX]");

        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSSxx")
                .withZone(ZoneOffset.UTC);

        DateTimeFormatter dtf = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSS[xx][XX]");

        return Mono.just(new FunctionOutput(
                List.of(EventResult.outputOf(
                        Map.of(OUTPUT, new JsonPrimitive((ZonedDateTime.parse(firstDate, dtf).toEpochSecond()
                                - ZonedDateTime.parse(secondDate, dtf).toEpochSecond()) * 1000))))));
    }

}

File: kirun-js/src/engine/function/system/date/DifferenceOfTimestamps.ts
import { KIRuntimeException } from "../../../exception/KIRuntimeException";
import { Schema } from "../../../json/schema/Schema";
import { EventResult } from "../../../model/EventResult";
import { FunctionOutput } from "../../../model/FunctionOutput";
import { FunctionSignature } from "../../../model/FunctionSignature";
import { Parameter } from "../../../model/Parameter";
import { Namespaces } from "../../../namespaces/Namespaces";
import { FunctionExecutionParameters } from "../../../runtime/FunctionExecutionParameters";
import { ValidDateTimeUtil } from "../../../util/ValidDateTimeUtil";
import { AbstractFunction } from "../../AbstractFunction";

const DATE_ONE: string = "isoDateOne";
const DATE_TWO: string = "isoDateTwo";
const OUTPUT: string = "result";

const SIGNATURE: FunctionSignature = new FunctionSignature('DifferenceOfTimestamps')
    .setNamespace(Namespaces.DATE)
    .setParameters(
        new Map([
            [DATE_ONE, new Parameter(DATE_ONE, Schema.ofString(DATE_ONE).setRef(Namespaces.DATE+".timeStamp"))],
            [DATE_TWO, new Parameter(DATE_TWO, Schema.ofString(DATE_TWO).setRef(Namespaces.DATE+".timeStamp"))]
        ])
    );;

export class DifferenceOfTimestamps extends AbstractFunction {

    public getSignature(): FunctionSignature {
        return SIGNATURE;
    }

    protected async internalExecute(context: FunctionExecutionParameters): Promise<FunctionOutput> {

        const firstDate:string = context?.getArguments()?.get(DATE_ONE);
        const secondDate:string = context?.getArguments()?.get(DATE_TWO);

        if(!ValidDateTimeUtil.validate(firstDate) || !ValidDateTimeUtil.validate(secondDate)) 
            throw new KIRuntimeException("Please provide valid ISO date for both the given dates.");

        const fDate: Date = new Date(firstDate);
        const sDate: Date = new Date(secondDate);

        return new FunctionOutput([EventResult.outputOf(new Map([[OUTPUT, (sDate.getTime() - fDate.getTime())/60000]]))]);
    }

}